### PR TITLE
Automatically push and pull new databases in replicated servers

### PR DESCRIPTION
**Pull request by zachmu**

### CODE REVIEW

File: go/libraries/doltcore/sqle/database_provider.go [hunk 1]
```diff
@@ -177,13 +178,55 @@ func (p DoltDatabaseProvider) Database(ctx *sql.Context, name string) (db sql.Da
...
+
+	// TODO: error handling when not set
+	_, remoteUrlTemplate, _ := sql.SystemVariables.GetGlobal(dsess.ReplicationRemoteURLTemplate)
+	if remoteUrlTemplate == "" {
+		return nil
+	}
+
+	// TODO: url sanitize
+	// TODO: SQL identifiers aren't case sensitive, but URLs are, need a plan for this
+	remoteUrl := fmt.Sprintf(remoteUrlTemplate.(string), dbName)
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/database_provider.go:216]: Right now the URL template has a printf marker (%s) in it, but that's probably not what we want.

Maybe `$database$`? Thoughts?

File: go/libraries/doltcore/doltdb/commit_hooks.go [hunk 2]
```diff
@@ -16,42 +16,91 @@ package doltdb
...
+	if err != nil {
+		return err
+	}
+
+	ds, err = ph.destDB.GetDataset(ctx, rf.String())
+	if err != nil {
+		return err
+	}
+
+	_, err = ph.destDB.SetHead(ctx, ds, addr)
```
Comment (zachmu) [inline go/libraries/doltcore/doltdb/commit_hooks.go:94]: Without this call, the pull.Pull above succeeds but results in a manifest with a 0 value root, and I don't get why

**Review by reltuk — COMMENTED** [review]
A few comments. Generally looks good!

File: go/libraries/doltcore/doltdb/commit_hooks.go [hunk 3]
```diff
@@ -33,25 +34,74 @@ type PushOnWriteHook struct {
...
+	targetHash, ok := ds.MaybeHeadAddr()
+	if !ok {
+		return fmt.Errorf("dataset empty")
+	}
+
+	addr, ok := ds.MaybeHeadAddr()
+	if !ok {
+		_, err := ph.destDB.Delete(ctx, ds)
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (zachmu) [reply go/libraries/doltcore/doltdb/commit_hooks.go:73]: Yes, we replicate branch deletes. Fixing.

File: go/libraries/doltcore/doltdb/commit_hooks.go [hunk 4]
```diff
@@ -33,25 +34,74 @@ type PushOnWriteHook struct {
...
+	if err != nil {
+		return err
+	}
+
+	ds, err = ph.destDB.GetDataset(ctx, rf.String())
+	if err != nil {
+		return err
+	}
+
+	_, err = ph.destDB.SetHead(ctx, ds, addr)
```
[the remark this answers is not in the document]
Comment (zachmu) [reply go/libraries/doltcore/doltdb/commit_hooks.go:94]: Thanks for clearing it up!

File: go/libraries/doltcore/env/remotes.go [hunk 5]
```diff
@@ -476,12 +495,21 @@ func getAbsFileRemoteUrl(urlStr string, scheme string, fs filesys2.Filesys) (str
 	}
 
 	urlStr = strings.ReplaceAll(urlStr, `\`, "/")
-	if !strings.HasPrefix(urlStr, "/") {
+	if !strings.HasPrefix(urlStr, "/") && !isDriveLetterPath(urlStr) {
 		urlStr = "/" + urlStr
 	}
```
[the remark this answers is not in the document]
Comment (zachmu) [reply go/libraries/doltcore/env/remotes.go:500]: You might be right that this code is useless and was only ever running on windows. I'll try removing it and see where it gets us.

File: go/libraries/doltcore/sqle/database_provider.go [hunk 6]
```diff
@@ -292,7 +343,54 @@ func (p DoltDatabaseProvider) CreateDatabase(ctx *sql.Context, name string) erro
...
+		return nil
+	}
+
+	// TODO: url sanitize
+	remoteUrl := fmt.Sprintf(remoteUrlTemplate.(string), name)
+
+	// TODO: params for AWS, others that need them
+	r := env.NewRemote(remoteName, remoteUrl, nil)
+	// TODO: use the correct format here, should match created DB
+	err := r.Prepare(ctx, types.Format_Default, p.remoteDialer)
```
[the remark this answers is not in the document]
Comment (zachmu) [reply go/libraries/doltcore/sqle/database_provider.go:373]: Good catch

File: go/libraries/doltcore/sqle/database_provider.go (hunk 1 above)
Comment (zachmu) [reply go/libraries/doltcore/sqle/database_provider.go]: I went with {database}, thanks!

**Pull request merged**